component/src/main/java/org/ballerinalang/sap/nativeimpl/consumer/RFCConnectionHandler.java

```diff
@@ -0,0 +1,111 @@
+import org.ballerinalang.model.values.BError;
+import org.ballerinalang.model.values.BMap;
+import org.ballerinalang.model.values.BValue;
+import org.ballerinalang.sap.utils.SapUtils;
+import org.ballerinalang.util.codegen.ProgramFile;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.ballerinalang.sap.utils.SapConstants.CONSUMER_STRUCT_NAME;
+import static org.ballerinalang.sap.utils.SapConstants.RESOURCE_ON_MESSAGE;
+import static org.ballerinalang.sap.utils.SapConstants.SAP_MESSAGE_OBJECT;
+import static org.ballerinalang.sap.utils.SapConstants.SAP_NATIVE_PACKAGE;
+
+/**
+ * This class handles BAPI calls returned from the SAP gateway.
+ * <p>
+ * This class encapsulates each bapi/rfc request received.
+ * handling  and converting them to String
+ * </p>
+ */
+public class RFCConnectionHandler implements JCoServerFunctionHandler {
+
+    private static final Log log = LogFactory.getLog(RFCConnectionHandler.class);
+    private ResponseCallback callback;
+    private Map<String, Resource> sapService;
+
+    public RFCConnectionHandler(Map<String, Resource> sapService) {
+
+        this.sapService = sapService;
+        callback = new ResponseCallback();
+    }
+
+    /**
+     * handle bapi requests coming through SAP gateway
+     * @param jCoServerContext JCO Server environment configuration
+     * @param jCoFunction bAPI/rfc function being called
+     */
+    @Override
+    public void handleRequest(JCoServerContext jCoServerContext, JCoFunction jCoFunction) {
+
+        if (log.isDebugEnabled()) {
+            log.debug("New BAPI function call received");
+        }
+        Map<String, Object> properties = new HashMap<>();
+        Resource onMessageResource = sapService.get(RESOURCE_ON_MESSAGE);
+        ProgramFile programFile = onMessageResource.getResourceInfo().getPackageInfo().getProgramFile();
+        BMap<String, BValue> consumerStruct = BLangConnectorSPIUtil.createBStruct(programFile,
+                SAP_NATIVE_PACKAGE, CONSUMER_STRUCT_NAME);
+        Executor.submit(onMessageResource, callback, properties, null,
+                getSignatureParameters(jCoFunction, consumerStruct));
+    }
+
+    private BMap<String, BValue> getSignatureParameters(JCoFunction jCoFunction,
+                                                        BMap<String, BValue> consumerStruct) {
+
+        try {
+            log.debug("Starting a new BAPI process from the incoming request");
+            jCoFunction.getExportParameterList().setValue("ECHOTEXT",
+                    jCoFunction.getImportParameterList().getString("REQUTEXT"));
+            log.info("Request text: " + jCoFunction.getImportParameterList().getString("REQUTEXT"));
+            consumerStruct.addNativeData(SAP_MESSAGE_OBJECT, jCoFunction.getImportParameterList()
+                    .getString("REQUTEXT"));
+        } catch (Throwable e) {
+            SapUtils.invokeOnError(sapService, callback, e.getMessage());
```

Comment:
If an exception is thrown, we would call `invokeOnError`, but then the `consumerStruct` will be returned at L96 right? And then we will be dispatching to `onMessage` in L79. So we will be dispatching to both `onMessage` and `onError` even when an error occurs right? 

We should only dispatch to one of the two.